theano/gpuarray/ctc.py

```diff
@@ -0,0 +1,185 @@
+from theano.gof import local_optimizer
+from theano.tensor.opt import register_canonicalize
+from theano.tensor.opt import register_stabilize
+
+import os
+import os.path
+from . import pygpu
+
+ctc_enabled = config.ctc.enabled
+
+
+class GpuConnectionistTemporalClassification(gof.COp):
+    """
+    GPU wrapper for Baidu CTC loss function.
+
+    Parameters
+    ----------
+    compute_grad
+        If set to True, enables the computation of gradients of the CTC loss function.
+    """
+    __props__ = ('compute_grad',)
+
+    func_file = "./c_code/ctc_wrapper.c"
+    func_name = "APPLY_SPECIFIC(ctc_cost_gpu)"
+
+    params_type = gpu_context_type
+
+    def __init__(self, compute_grad=True):
+        if not compute_grad:
+            self.func_name = "APPLY_SPECIFIC(ctc_cost_gpu_no_grad)"
+        self.compute_grad = compute_grad
+
+        gof.COp.__init__(self, self.func_file, self.func_name)
+
+        if config.ctc.root == "":
+            raise ValueError('ctc.root variable is not set, please set it '
+                             'to the root directory of the CTC library in '
+                             'your system.')
+
+    def c_lib_dirs(self):
+        dirs = []
+        if ctc_enabled:
+            # Find the directory that contains libwarpctc.so
+            lib_found = False
+            for lib_dir in ["build", "lib", "lib64"]:
+                lib_path = os.path.join(config.ctc.root, lib_dir)
+                if os.path.isdir(lib_path) and os.path.exists(lib_path):
+                    lib_found = os.path.exists(os.path.join(lib_path, "libwarpctc.so"))
+                    if lib_found:
+                        dirs.append(lib_path)
+                        break
+
+            if not lib_found:
+                raise RuntimeError('libwarpctc.so could not be found. ',
+                                   'Please check the config.ctc.root variable.')
+        return dirs
+
+    def c_libraries(self):
+        return ["warpctc", "gpuarray"]
+
+    def c_header_dirs(self):
+        dirs = [os.path.dirname(__file__), pygpu.get_include()]
+        if ctc_enabled:
+            # We assume here that the header is available at the include directory
+            # of the CTC root directory.
+            dirs.append(os.path.join(config.ctc.root, "include"))
+        return dirs
+
+    def c_headers(self):
+        return ['ctc.h', 'numpy_compat.h', 'gpuarray_helper.h', 'gpuarray/types.h',
+                'gpuarray_api.h', 'gpuarray/array.h', 'gpuarray/util.h', 'gpuarray/extension.h']
+
+    def get_params(self, node):
+        return node.inputs[0].type.context
+
+    def make_node(self, activations, labels, input_lengths):
+        if not ctc_enabled:
+            raise RuntimeError('Baidu CTC is not enabled and '
+                               'GpuConnectionistTemporalClassification Op '
+                               'can not be constructed.')
+
+        context = infer_context_name(activations, labels, input_lengths)
```

Comment:
Please change the name of the variable to `context_name` to avoid confusion.

Also, if you want labels and input_lengths as CPU variables, don't look at them to choose a context.